src/pass/zero_elimination.cc

```diff
@@ -0,0 +1,1719 @@
+    if (imm && imm->value > 0) {
+      auto it = expr_to_vars_.find({op->a.get(), imm->value});
+      if (it != expr_to_vars_.end()) {
+        return it->second.first;
+      }
+
+      Expr mutated_a = Mutate(op->a);
+      if (auto var_pair_opt = AddNewVarPair(op->a, mutated_a, imm->value)) {
+        return var_pair_opt.value().first;
+      } else {
+        return Div::make(mutated_a, Mutate(op->b));
+      }
+    }
+
+    return Div::make(Mutate(op->a), Mutate(op->b));
+  }
+
+  virtual Expr Mutate_(const Mod* op, const Expr& e) {
+    const IntImm* imm = op->b.as<IntImm>();
+    if (imm && imm->value > 0) {
+      auto it = expr_to_vars_.find({op->a.get(), imm->value});
+      if (it != expr_to_vars_.end()) {
+        return it->second.second;
+      }
+
+      Expr mutated_a = Mutate(op->a);
+      if (auto var_pair_opt = AddNewVarPair(op->a, mutated_a, imm->value)) {
+        return var_pair_opt.value().second;
+      } else {
+        return Mod::make(mutated_a, Mutate(op->b));
+      }
+    }
+
+    return Mod::make(Mutate(op->a), Mutate(op->b));
+  }
+
+ private:
+  dmlc::optional<std::pair<Var, Var>> AddNewVarPair(const Expr& e, const Expr& mut, int64_t val) {
+    using tresult = dmlc::optional<std::pair<Var, Var>>;
+
+    Expr val_e = make_const(e.type(), val);
+    idx_ += 1;
+
+    std::unordered_map<const Variable*, IntSet> var_intsets;
+    for (const auto& p : ranges) {
+      var_intsets[p.first.get()] = IntSet::range(p.second);
+    }
+
+    Range div_range = EvalSet(mut / val_e, var_intsets).cover_range(Range());
+    Range mod_range = EvalSet(mut % val_e, var_intsets).cover_range(Range());
+
+    if (!div_range.get() || !mod_range.get()) {
+      LOG(WARNING) << "EliminateDivMod: won't eliminate div or mod of expr " << e
+                   << "  because its bounds cannot be inferred";
+      return tresult();
+    }
+
+    auto div = Var("div" + std::to_string(idx_), e.type());
+    auto mod = Var("mod" + std::to_string(idx_), e.type());
+
+    new_variables.push_back(div);
+    new_variables.push_back(mod);
+
+    substitution.Set(div, mut / val_e);
+    substitution.Set(mod, mut % val_e);
+
+    ranges.Set(div, div_range);
+    ranges.Set(mod, mod_range);
+
+    conditions.push_back(mut == div*val_e + mod);
+
+    if (!CanProve(mod_range->extent <= val_e)) {
+      LOG(WARNING) << "EliminateDivMod: cannot fully eliminate div or mod of expr " << e
+                   << "  (probably it may change its sign)";
+      // We cannot prove that mod is unique, so add additional condition
+      conditions.push_back(Select::make(e >= 0, mod >= 0, mod <= 0));
+    }
+
+    auto p = std::make_pair(div, mod);
+    expr_to_vars_[{e.get(), val}] = p;
+    return tresult(p);
+  }
+
+  int idx_{0};
+  std::map<std::pair<const HalideIR::Internal::IRNode*, int64_t>, std::pair<Var, Var>>
+    expr_to_vars_;
+};
+
+// replace every subexpr of the form e/const and e % const with a new variable
```

Comment:
If we have `x/3 + (x/3)*(y/3)`, will it be transformed to `v1 + v1/v2` or `v1+v2/v3`? Please clarify this in the comments.

Reply:
Modified the class so that the answer to this question is now "yes".